dev-docs/source/DesignDocumentGuides.rst

```diff
@@ -10,39 +10,96 @@ Design Document Guidelines
 they often add unnecessary complexity if not done with care. These
-guidelines are to help you generate effective and simple design
+guidelines are to help when and how to generate effective and simple design
 documents.
 
-Before starting, all Mantid design documents must be written in
-`markdown <http://en.wikipedia.org/wiki/Markdown>`__ format with the md
-extension. After approval from the the Mantid Project :ref:`TSC`, they should be stored
-`here <https://github.com/mantidproject/documents/tree/master/Design>`__
-or in one of the subdirectories of that directory. This page covers what
-happens between those points.
+When to write a design document
+###############################
 
-Guidelines on Sections
-######################
+This section provides some guidelines on when to produce a design document.
+Note that, it is not about when to do a design and when not.
+One should always do design work before coding, no matter the scope and the nature of the feature.
+These guidelines are to answer in which cases the design must be a *public*, *written* and *persistent* document, rather than a sketch on a white-board in a stand-up meeting that is wiped-out after it has been turned into code.
+Just as coding, design is an iterative and collaborative process.
+Having the document written allows for discussion and iterations with peers before the first line of code is written.
+Besides, the written document facilitates the pull request review process.
+Having the design already approved, the reviewer will not need to do a design review with the code review, but will just need to make sure that the implementation matches the previously agreed design.
+Finally, once written, the document can also serve as (or easily be turned into) a developer documentation for future reference.
+
+Not every feature requires a design document, as the benefit it adds must be gaged with the time and effort that goes into it.
+Bug fixes, small extensions to the current functionalities, as well as concrete, single-class plugins, such as algorithms, do not need a document.
+In the contrary, if the feature is large-scope, brand-new, many-class, including abstractions, and/or touches or refactors the internals of the core framework, it will most likely need a document.
+Below are a couple of metrics that could help a developer decide whether or not the feature falls into that category.
+
+- Number of classes and components.
+
+  If the number of classes is more than just a few, the document is relevant in order to describe the relations between them.
+
+- Fraction of abstract classes and interfaces in the design.
+
+  Abstractions and APIs are hard to change, once in. So whenever the feature concerns creations of new abstraction layers and APIs, a wider discussion on the design is needed, hence the necessity of the document.
+
+- Fraction of files and components the feature will be touching.
+
+  Single file edits are less likely to require a document.
```

Comment:
Can we expand on this? Is there a point where developers really must considering a written design document such as the number of files/components is greater than x or the percentage of the code base maybe? Perhaps we could also say that for cases inbetween a single file and the point at which you definitely need to do this that they need to be judged on a case by case basis? We would need to be clear what the criteria for judgement would be though.

Reply:
It is hard to find the appropriate thresholds. That's why these stay as guidelines, otherwise they will be just rules. I could elaborate on the extreme cases, such as one file, one class, or the other extreme. But in reality, most of the features will fall into the gray zone. Hence, the purpose is to combine all this metrics and use some overall judgement and/or advice from seniors to decide.

Reply:
It's a really difficult one. I appreciate we are not trying to set rules here but we're potentially straying into the territory of guidance so vague it's not worth providing it. Perhaps a sentence about anything larger than a single file edit needs to be thought about and a reminder to use their judgement and/or seek advice from seniors. Then we're not setting rules but we're not being far too vague either.